sympy/polys/rootconditions.py

```diff
@@ -0,0 +1,265 @@
+
+    qs = [p[1] * qi for qi in p]
+    for i in range(1, len(p), 2):
+        qs[i - 1] -= p[i] * p[0]
+    qs = qs[1:]
+
+    cond = p[0] * p[1]
+
+    # This check is really important. It stops the recursion immediately and
+    # avoid a possible infinite while loop in _clear_cond_poly.
+    if cond.is_zero:
+        return [K.zero]
+
+    cond = _clear_cond_poly(cond, previous_cond, K)
+
+    # These checks are also important, because adding 1 or -1 to the previous
+    # conditions will lead to an infinite loop in the next _clear_cond_poly.
+    if cond == K(-1):
+        return [K(-1)]
+
+    if not cond.is_one:
+        previous_cond.append(cond)
+
+    return _rec_dup_routh_hurwitz_poly(qs, previous_cond, K)
+
+def _clear_cond_poly(cond: PolyElement[Er], previous_cond, K):
+    # Divide out factors known to be positive from previous conditions.
+
+    # There are not controls in that functions on cond and previous_cond,
+    # we assume that at this point, there are no zeroes, ones and negative ones
+    # in previous_cond.
+    # If there are, there will be an infinite while loop.
+    for c in previous_cond:
+        cond_quo, r = K.div(cond, c)
+        while not r:
+            cond = cond_quo
+            cond_quo, r = K.div(cond, c)
+
+    # Remove factors of even degree and reduce odd degree factors
+    common_factor, facs_m = cond.sqf_list()
+
+    cond = K(prod([fac for fac, m in facs_m if m % 2 == 1]))
+    if common_factor < 0:
+        cond = cond * K(-1)
+
+    return cond
+
+
+def _dup_routh_hurwitz_exraw(p: list[Er]) -> list[Er]:
+    if all(c.is_Number for c in p):
+        return _calc_conditions_div(p)
+
+    return _calc_conditions_no_div(p)
+
+def _calc_conditions_div(p: list[Er]) -> list[Er]:
+    """
+    Stability check with divisions, used for numeric cases.
+
+    """
+    if len(p) < 2:
+        return [EXRAW.one]
+
+    if (p[0]*p[1]).is_nonpositive:
+        return [EXRAW(-1)]
+
+    if len(p) == 2:
+        return [p[0]*p[1]]
+
+    qs = p.copy()
+    for i in range(1, len(p), 2):
+        qs[i - 1] -= p[i] * p[0] / p[1]
+    qs = qs[1:]
+
+    return [p[0] * p[1]] + _calc_conditions_div(qs)
+
+def _calc_conditions_no_div(p: list[Er]) -> list[Er]:
+    """
+    Stability check without divisions, used for EXRAW.
+
+    """
+    return _rec_calc_conditions_no_div(p, [])
+
+def _rec_calc_conditions_no_div(p: list[Er],
+                                previous_cond: list[set]) -> list[Er]:
+    if len(p) < 2:
+        return [EXRAW.one]
+
+    if len(p) == 2:
+        return [_clear_cond_exraw(p[0]*p[1], previous_cond)]
+
+    qs = [p[1] * qi for qi in p]
+    for i in range(1, len(p), 2):
+        qs[i - 1] -= p[i] * p[0]
+    qs = qs[1:]
+
+    cond: Mul = _clear_cond_exraw(p[0] * p[1], previous_cond)
+    return [cond] + _rec_calc_conditions_no_div(qs, previous_cond)
+
+def _clear_cond_exraw(cond: Mul, previous_cond: list[set]) -> Mul:
```

Comment:
The annotationw here should be Expr rather than Mul because in general `Mul(*args)` creates an Expr rather than a Mul.